### PR TITLE
Make `dafny verify` more consistent

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
- Enable using the `--filter-[position|symbol]` options together with `--log-format` and with `--verification-coverage-report`
- Let `--log-format` and `--verification-coverage-report` collect data in a streaming fashion, so memory usage doesn't increase during verification.
- Make architectural changes that would enable `--log-format` and `--verification-coverage-report` to report based on Dafny declarations instead of Boogie implementations.

### How has this been tested?
- Added tests for the JSON and text verification logger when used with the old CLI, since that code was modified and the existing tests only test the new CLI.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyDriver/Commands/VerifyCommand.cs [hunk 1]
```diff
@@ -28,24 +27,22 @@ public static class VerifyCommand {
     foreach (var option in Options) {
       result.AddOption(option);
     }
-    DafnyNewCli.SetHandlerUsingDafnyOptionsContinuation(result, async (options, _) => {
-      if (options.Get(CommonOptionBag.VerificationCoverageReport) != null) {
-        options.TrackVerificationCoverage = true;
-      }
-
-      if (options.Get(CommonOptionBag.VerificationLogFormat).Any() || options.Get(CommonOptionBag.VerificationCoverageReport) != null) {
```
Comment (keyboardDrummer) [inline Source/DafnyDriver/Commands/VerifyCommand.cs:36]: Note this `if` is gone in `HandleVerification`, so `dafny verify` no longer switches to using `SynchronousCliCompilation` when one of those options is used, which is the main contribution of this PR.

**Comment by atomb** [comment]
>     * `--log-format` and `--verification-coverage-report` are no longer supported when using the old CLI. I could copy paste some code from before this PR to still allow them to work, but that code won't be tested and the behavior of the old CLI for those option will likely diverge from the new CLI.

The old-style `/verificationLogger` flag is still in active use, and I wouldn't want us to remove it until I was very confident that the new CLI could replace it in every situation.

**Comment by keyboardDrummer** [comment]
> > ```
> > * `--log-format` and `--verification-coverage-report` are no longer supported when using the old CLI. I could copy paste some code from before this PR to still allow them to work, but that code won't be tested and the behavior of the old CLI for those option will likely diverge from the new CLI.
> > ```
> 
> The old-style `/verificationLogger` flag is still in active use, and I wouldn't want us to remove it until I was very confident that the new CLI could replace it in every situation.

Brought back the feature for the old CLI

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/github-issue-2563.dfy [hunk 2]
```diff
@@ -1,2 +1,2 @@
-// RUN: %exits-with 1 %dafny /compile:0 '/verificationLogger:csv;log.csv' "%s" > "%t"
```
Comment (atomb) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/github-issue-2563.dfy:1]: Could we leave in both lines to have some assurance that they produce the same results?

**Review by atomb — APPROVED** [review]
This looks very nice.

It seems like this would make it easier to add streaming output indicating verification status. Is that right? I tended to use Boogie's `/trace` flag for this, but it's not at the right level of abstraction, and no longer produces output on verification completion when used with `dafny verify`. Having a Dafny-centric option that would print a message summarizing each `VerificationRunResult` to the console as it became available would be very nice.

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/github-issue-2563.dfy (hunk 2 above)
Comment (keyboardDrummer) [reply Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/github-issue-2563.dfy:1]: They don't produce the same results. You can see from the diff in the expect file that the error message is slightly different. There's already various small differences in reporting from the new and old CLI, and generally we only test the new CLI in that case. If you want I can test both here but it's a complicated topic.

**Comment by keyboardDrummer** [comment]
> It seems like this would make it easier to add streaming output indicating verification status. Is that right?

Definitely. The new architecture uses Boogie as a library that outputs events, among which a single assertion batch having been completed, and Dafny is free to emit output based on those.

We also have the option of enhancing the Dafny CLI to detect whether it's in a terminal that supports escape codes, and if so show dynamic output such as a progress bar. We could use this .NET library to help us create that: https://spectreconsole.net/

**Pull request merged**